oslo/torch/nn/parallel/distributed/pipeline_parallel/p2p.py

```diff
@@ -0,0 +1,132 @@
+        ctx.save_for_backward(module_rank, module_rank_parent)
+        return _pp_pre_fwd_p2p_com(input_, module_rank, module_rank_parent)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        module_rank, module_rank_parent, = ctx.saved_tensors
+        return _pp_post_bwd_p2p_com(grad_output, module_rank, module_rank_parent)
+    
+    
+### PP post fwd p2p com
+
+def _pp_post_fwd_p2p_com(output_, module_rank, module_rank_parent):
+    rank = dist.get_rank()
+    if output_.device.index != module_rank_parent:
+        if module_rank == rank: # if output_ is on our rank we send
+            request = dist.isend(tensor=output_, dst=module_rank_parent)
+            request.wait()
+        elif module_rank_parent == rank: # if the parent rank is our rank we receive output_
+            output_buffer = torch.empty(output_.shape, device=rank)
+            request = dist.irecv(tensor=output_buffer, src=module_rank)
+            request.wait()
+            return output_buffer
+    else: # data_out and module are on the same rank
+        return output_
+
+
+def _pp_pre_bwd_p2p_com(output_grad, module_rank, module_rank_parent):
+    rank = dist.get_rank()
+    if output_grad.device.index != module_rank:
+        if data_out_grad.device.index == rank: # if output_grad is on our rank we send
+            request = dist.isend(tensor=output_grad, dst=module_rank)
+            request.wait()
+        elif module_rank == rank: # if the module rank is our rank we receive output_grad
+            output_grad_buffer = torch.empty(output_grad.shape, device=rank)
+            request = dist.irecv(tensor=output_grad_buffer, src=output_grad.device.index)
+            request.wait()
+            return data_out_grad_buffer
+    else: # data_out_grad and module are on the same rank
+        return output_grad
+
+
+class PPPostFwdP2PCom(torch.autograd.Function):
+    
+    @staticmethod
+    def forward(ctx, input_, module_rank=None, module_rank_parent=None):
+        ctx.save_for_backward(module_rank, module_rank_parent)
+        return _pp_post_fwd_p2p_com(input_, module_rank, module_rank_parent)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        module_rank, module_rank_parent, = ctx.saved_tensors
+        return _pp_pre_bwd_p2p_com(grad_output, module_rank, module_rank_parent)
+
+
+### nn.Module wrapper
+
+class PPModuleWrapper(nn.Module):
+    def __init__(self, module):
+        super().__init__()
+        self.module = module
+        self.rank = module.rank
+        if hasattr(module, "pp_rank_parent"):
+            self.rank_parent = module.pp_rank_parent
+        else:
+            self.rank_parent = None
+        self.pre_com = PPPreFwdP2PCom(self.rank, self.rank_parent).apply
+        self.post_com = PPPostFwdP2PCom(self.rank, self.rank_parent).apply
+        
+    def forward(self, x):
+        x = self.pre_com(x)
+        x = self.module(x)
+        x = self.post_com(x)
+        return x
+
+
+def wrap_nn_modules(m):
```

Comment:
@MicPie @Quentin-Anthony  Did you test this?

Reply:
Yes, local test works as shown below:
```
> model = CustomModule(); model
CustomModule(
  (module1): Linear(in_features=8, out_features=8, bias=True)
  (module2): Sequential(
    (0): Linear(in_features=8, out_features=8, bias=True)
    (1): Sequential(
      (0): Linear(in_features=8, out_features=8, bias=True)
      (1): Linear(in_features=8, out_features=8, bias=True)
    )
    (2): Linear(in_features=8, out_features=8, bias=True)
  )
)

> wrap_nn_modules(model); model
CustomModule(
  (module1): PPModuleWrapper(
    (module): Linear(in_features=8, out_features=8, bias=True)
  )
  (module2): PPModuleWrapper(
    (module): Sequential(
      (0): PPModuleWrapper(
        (module): Linear(in_features=8, out_features=8, bias=True)
      )
      (1): PPModuleWrapper(
        (module): Sequential(
          (0): PPModuleWrapper(
            (module): Linear(in_features=8, out_features=8, bias=True)
          )
          (1): PPModuleWrapper(
            (module): Linear(in_features=8, out_features=8, bias=True)
          )
        )
      )
      (2): PPModuleWrapper(
        (module): Linear(in_features=8, out_features=8, bias=True)
      )
    )
  )
)
```

Reply:
Let me update `oslo/torch/nn/parallel/distributed/pipeline_parallel/test_model.py` to have that included.

Reply:
Working test pushed.